### ISSUE TITLE
Help !

### ISSUE THREAD

**Issue by Maatary** [TimWSpence/cats-stm#299]
Hi,

I can't find a lot of documentation or post online about this module of cats which i find amazing. Where can i find good material to learn from it. 

I have a use case where i am relying on a very old Java api, which is all about mutation. I want to isolate it and get back a sense of referential transparency. The all api, is all about mutating value. It is an api to mange RDF model. 

Hence I was thinking about finding an abstraction other than just IO, which would help me write some good referentially transparent code and track the mutation

any advise ?

**Comment by Maatary** [comment]
In short, I wonder if cats-stm, works with mutable value and not just mutable references ?

**Comment by TimWSpence** [comment]
Hi @Maatary! I'm assuming you've looked at https://timwspence.github.io/cats-stm? If not then that's the place to start. If you have and it hasn't helped then any feedback on what's missing/how it could be improved would be immensely helpful!!

It does sound like `IO` is what you want for your use-case however (taking a side-effectful API and making it referentially transparent). Was there a reason why you didn't want to use that?

That's actually a very interesting question about mutable values. STM was definitely designed to work with immutable data but it _might_ work with mutable data as the underlying reference to the value is only updated when have an exclusive lock and therefore know that the update will succeed. I'll need to go away and think about that one. Thanks for asking! :)

**Comment by TimWSpence** [comment]
I've thought about it and I'm afraid cats-stm fundamentally won't work with mutable values. https://github.com/TimWSpence/cats-stm/blob/master/core/src/main/scala/io/github/timwspence/cats/stm/STMLike.scala#L468 is the exact place where it becomes a problem. Essentially when committing a txn, cats-stm does an initial (potentially stale) read of the value in a `TVar` and then subsequent modifications just touch that local copy. When we come to commit, we acquire a lock and check that the initial value we read is not stale (which would require a retry). But if the value is mutable then we would be mutating the actual value stored in the `TVar` instead of a local copy.

**Comment by TimWSpence** [comment]
@Maatary can I help you with anything else? Otherwise I'll close this issue

**Comment by Maatary** [comment]
@TimWSpence sorry for the late response. Got swamped at work but also doing my own learning. 

So yes you were right, perfect case for IO. I did not know well IO and cats effects. This use case has been my vehicle to learn all about cats effects and al. I now know how to deal with mutability ( IO Action or Action that affect the state of the world) in a referentially transparent manner. 

Me looking at STM or Ref before that, was due to a fundamental lack of comprehension of the interplay between the notion of side effect and referential transparency. I was focus on a limited understanding of referential transparency i.e. "Same input give Same output". Because of that, i thought about garanteeing that with concurrency construct such as Ref of STM. That is, locking the input, would garantee the same output. I finally understood that is was the wrong way to fundamentally think about the all problem. 

First the Ref or STM would not help because the underlying structure is mutable. Second, the issue was to understanding that i simply needed to represent my IO Action as a value to recover referential transparency and therefore local reasoning, thanks to the region of sharing. This is opposed to perform the IO Action directly (as a side effect) in the middle of the program such that i would need to keep in my head the new state of the world, in the subsequent part of the program. 

The all context of this is here: https://stackoverflow.com/questions/66172722/using-scala-cats-io-type-to-encapsulate-a-mutable-java-library

This is what sparked my journey in understanding all of this. Your answer above, helped guide my study and confirm ultimately my understanding of this. So thank you.
